### DISCUSSION TITLE
Preferred Way to Run Local Tests

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvista#2539] [Q&A] [answered]
Should developers use [github actions act](https://github.com/nektos/act) to run tests locally in an isolated environment prior to a PR? It uses Docker, so Windows developers can run Windows and Linux tests locally, but not Mac tests. Has anyone used this? I'm looking for an elegant solution to test locally for all supported OS's without having to push multiple times to the repo (it clutters the history and uses up runner time).

**Comment by banesullivan** [comment, marked as answer]
*GitHub actions act* seems like a super useful utility -- I will definitely be checking that out for some other projects with simpler CI configurations.

I'm not sure if PyVista is the best candidate for that utility as we've put a lot of work into caching and optimizing those actions specifically for running in CI. With that said, I'd definitely like to hear about your experience using *GitHub actions act* for PyVista's tests if you give it a try!

>  without having to push multiple times to the repo (it clutters the history and uses up runner time)

From my perspective, this isn't a big deal. We squash and merge all PRs so that there is only one commit per PR making any debugging commits on that branch disappear. Sometimes I like to see that debugging history on the branch as it captures why some small thing broke the windows tests which is useful context for me being a non-windows user.

As far as runner time goes, GitHub's OS tier is very generous and I don't think we've come anywhere close to our limit operating this way... so I'm inclined to say, "*if it ain't broke, don't fix it*"

What I will personally do: run `pytest -n 10` locally and push debugging commits to feature branches and let the CI handle testing windows (also necessary for me as I am on Mac and Linux machines)

**Reply by tkoyama010** [reply]
I am a fan of the PyVista community's squashing PR policy. They erase our terrible history of trial and error. Sometimes I don't even run local tests and let GitHub CI do everything. It makes development like a game of turning red into green.

**Reply by adam-grant-hendry** [reply]
Thank you both. I was unaware we squash committed. This is good to know.

I also agree: it's better to use the CI system (esp. since I can only test on one system).

**Reply by adam-grant-hendry** [reply]
Github Actions act might be more useful when first designing the CI/CD system. I would use it to write an initial `.github/worflow.yml`.
